src/main/java/com/google/firebase/database/core/JvmAuthTokenProvider.java

```diff
@@ -35,51 +35,47 @@
   private final ScheduledExecutorService executorService;
   private final FirebaseApp firebaseApp;
 
-  public JvmAuthTokenProvider(FirebaseApp firebaseApp, ScheduledExecutorService executorService) {
+  JvmAuthTokenProvider(FirebaseApp firebaseApp, ScheduledExecutorService executorService) {
     this.executorService = executorService;
     this.firebaseApp = firebaseApp;
   }
 
-  private static String wrapOAuthToken(FirebaseApp firebaseApp, GetTokenResult result) {
-    String oauthToken = result.getToken();
-    if (oauthToken == null) {
+  private static String wrapOAuthToken(FirebaseApp firebaseApp, AccessToken result) {
+    if (result == null) {
       // This shouldn't happen in the actual production SDK, but can happen in tests.
       return null;
-    } else {
-      Map<String, Object> authVariable = firebaseApp.getOptions().getDatabaseAuthVariableOverride();
-      GAuthToken googleAuthToken = new GAuthToken(oauthToken, authVariable);
-      return googleAuthToken.serializeToString();
     }
+
+    Map<String, Object> authVariable = firebaseApp.getOptions().getDatabaseAuthVariableOverride();
+    GAuthToken googleAuthToken = new GAuthToken(result.getTokenValue(), authVariable);
+    return googleAuthToken.serializeToString();
   }
 
   @Override
   public void getToken(boolean forceRefresh, final GetTokenCompletionListener listener) {
-    ImplFirebaseTrampolines.getToken(firebaseApp, forceRefresh)
-        .addOnCompleteListener(
-            this.executorService,
-            new OnCompleteListener<GetTokenResult>() {
-              @Override
-              public void onComplete(@NonNull Task<GetTokenResult> task) {
-                if (task.isSuccessful()) {
-                  listener.onSuccess(wrapOAuthToken(firebaseApp, task.getResult()));
-                } else {
-                  listener.onError(task.getException().toString());
-                }
-              }
-            });
+    GoogleCredentials credentials = ImplFirebaseTrampolines.getCredentials(firebaseApp);
+    try {
+      if (forceRefresh) {
+        credentials.refresh();
+      } else {
+        // getRequestMetadata() will refresh only if the credentials instance does not already
+        // have a valid (unexpired) OAuth2 token. GoogleCredentials does not provide a clearly
+        // labeled method for conditionally refreshing the underlying token. getRequestMetadata()
+        // is the closest thing to what we need.
+        credentials.getRequestMetadata();
```

Comment:
Thanks for adding the comment. I now understand what's going on a lot better and I'd actually suggest you change this again (sorry) so that we /always/ call getRequestMetadata() before calling getAccessToken() [this should be harmless in the case that we just called .refresh() since it will detect it doesn't need to refresh again] and then change the comment to something like:

// The typical way to use a GoogleCredentials object is to call getRequestMetadata() and include the metadata in your request. Since we are accessing the token directly via getAccessToken(), we must first call getRequestMetadata() to ensure the token is available (refreshed if necessary).

And while you're at it, I'd explicitly do "String accessToken = credentials.getAccessToken();" rather than embed it inside the listener.onSuccess() call. It's important enough to warrant its own line of code. :-)

Reply:
Done